dimagi/commcare-hq
# [SAAS-14520] Add conditionally_location_safe decorator to scheduling report views

## Product Description
This change allows location restricted web users to access views related to scheduling the emailing of saved reports. It also exposes views that are related to of saved and scheduled reports.

These changes are available behind a newly introduced feature flag - `location_restricted_scheduled_reports`

List of views for which `location_safe` decorator is added that checks for flag activation before exposing the view to location restricted web users:

| **#** 	| **View**                   	| **View's role**                                                                      	|
|-------	|----------------------------	|--------------------------------------------------------------------------------------	|
| 1     	| reports_home               	| To expose the base URL for reports home                                              	|
| 2     	| delete_config              	| To delete the saved report config                                                    	|
| 3     	| ScheduledReportsView       	| To expose the UI for Scheduled Reports                                               	|
| 4     	| delete_scheduled_report    	| To delete the scheduled report that may have one or more saved report config         	|
| 5     	| send_test_scheduled_report 	| To send the scheduled report via email                                               	|
| 6     	| view_scheduled_report      	| To view the content of scheduled report in current view without sending it via email 	|

## Technical Summary
https://dimagi-dev.atlassian.net/browse/SAAS-14520

Whitelist above listed views by adding `location_safe` decorator to their class/function. This makes the views available to the location restricted web users, which  currently are not allowed access, with views being blacklisted by default.

## Feature Flag
New feature flag - `location_restricted_scheduled_reports`

## Safety Assurance

### Safety story
Manual testing on staging
https://docs.google.com/document/d/1WrXtmQ5sY9T-alBkzmNH_GTVo1QURwfd17EqkW7Tqn8/edit

QA testing on staging
https://docs.google.com/document/d/1i7fCj_SxmDHfTKWxMS3KLCsNl0QFGc0haiwffuieq10/edit

As part of testing, a bug was found, for which fix is given as described below:
Fix added for scenario where users were still able to see data of all the locations if initially they were granted 'all locations' permission and later revoked. This is done by updating the filters that are used in SQL query context creation.

Since each of the exposed view is put behind the conditional check, these are available to location restricted web users only if the relevant feature flag is enabled for the domain. Along with it, extensive dev and QA testing is performed to cover different scenarios related to permissions and roles. These tests are described in the above links.

### Automated test coverage
Unit tests

### QA Plan
https://dimagi-dev.atlassian.net/browse/QA-5126


### Migrations
NA

### Rollback instructions

### Labels & Review
